### ISSUE TITLE
Keras custom loss function considering previous prediction

### ISSUE THREAD

**Issue by minsung-k** [keras-team/keras#19004]
Hi, I want to know how the custom loss function is called in epoch to consider previous prediction. Please check and let me know my method is right for that purpose.

Here's the toy dataset code.
`
# dataset
import keras
from keras import layers
import numpy as np

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
x_train = np.reshape(x_train, (-1, 784))
x_test = np.reshape(x_test, (-1, 784))

x_train = x_train[:100]
x_test = x_test[:50]
x_valid = x_test[50:100]

y1 = [*range(25, 0, -1)]
y2 = [*range(25, 0, -1)]
y3 = [*range(25, 0, -1)]
y4 = [*range(25, 0, -1)]

y5 = [*range(25, 0, -1)]
y6 = [*range(25, 0, -1)]

y7 = [*range(25, 0, -1)]
y8 = [*range(25, 0, -1)]

y_train = np.array(y1+y2+y3+y4)
y_test = np.array( y5+y6)
y_valid = np.array( y7+y8)
`
Let assume dataset is combined with 8 different dataset, all size is 25. I want to train my model that current output is smaller than the previous output since labels keep decreasing.

Here's the model.

`
inputs = keras.Input(shape=(784,), name="digits")
x1 = layers.Dense(64, activation="relu")(inputs)
x2 = layers.Dense(64, activation="relu")(x1)
outputs = layers.Dense(10, name="predictions")(x2)
model = keras.Model(inputs=inputs, outputs=outputs)

# Instantiate an optimizer.
optimizer =  tf.keras.optimizers.SGD(learning_rate=1e-3)

# compile
model.compile(optimizer = "adam", loss = model_loss(list_previous_label, list_true_label) )

# fit 
history  = model.fit(x_train, y_train, batch_size=25, epochs=10, verbose =2)
`

Here's the custom loss function.

`
# custom loss function

def model_loss(list_previous_label, list_true_label):

    list_previous_label = [100]
    list_true_label = [100]

    @tf.function
    def my_loss(y_true,y_pred):

        y_pred = tf.cast(y_pred, tf.float64)
        y_true = tf.cast(y_true, tf.float64)
        
        previous_prediction = tf.cast(list_previous_label[-1], tf.float64)
        previous_true = tf.cast(list_true_label[-1], tf.float64)
        
        violation_term = tf.cast(tf.math.maximum( (y_pred - previous_prediction) , 0), tf.float64)
          
        list_true_label.append(y_true)
        list_previous_label.append(y_pred)
        
        #print(violation_term)
        if previous_true < y_true:
            return tf.reduce_mean(tf.square(y_true - y_pred), axis=-1)
        
        else: return tf.reduce_mean(tf.square(y_true - y_pred), axis=-1) + (violation_term ** 2)
    return my_loss
`

My questions are 
1.does list_previous_label  and list_true_label save only one of 8 dataset and reset not to save when finish one dataset and initiate the next dataset? Since the last label of one dataset is always smaller than the next dataset's the first (0 vs 25). If not, how should I build my custom loss function? For that,  I set my batch size as 25.

2. Does list_previous_label  and list_true_label  can save each prediction? I cannot see how it works in the loss function.

Appreciate your consideration.

Best,
Min

**Comment by SuryanarayanaY** [comment]
Hi @minsung-k ,

The custom loss should have signature of `loss = fn(y_true, y_pred)` . Please find the documentation source below.

> loss: Loss function. May be a string (name of loss function), or a keras.losses.Loss instance. See keras.losses. A loss function is any callable with the signature loss = fn(y_true, y_pred), where y_true are the ground truth values, and y_pred are the model's predictions. y_true should have shape (batch_size, d0, .. dN) (except in the case of sparse loss functions such as sparse categorical crossentropy which expects integer arrays of shape (batch_size, d0, .. dN-1)). y_pred should have shape (batch_size, d0, .. dN). The loss function should return a float tensor.

It seems you are trying to pass `(list_previous_label, list_true_label)` which may not be possible.

Alternatively you can explore the custom training loop where you can configure your own training step using train_step method. You can refer the tutorials on how to override this using [1](https://keras.io/guides/custom_train_step_in_tensorflow/) & [2](https://www.tensorflow.org/tutorials/distribute/custom_training#define_the_loss_function).

Thanks!
